### PR TITLE
Reuse thing boxes when editing

### PR DESCRIPTION
**Pull request by ChrisRenfrow**
This pull request aims to address the enhancement put forward in #190.

The approach is as follows:
1. Have `/core/src/world/{npc,place}/view.rs` append the `uuid` of the Thing as a property of the `thing-box` element called `data-uuid`. (9b1391c217398f7a182fc8f324ba8701b2cac8c9)
2. In `/web/js/terminal.js`, intercept the output and check to see if the most recent thing-box element has a matching `data-uuid` property.
3. If it's a match, perform a diff of the two elements, flagging the changed fields with a class.
4. Use the class to add some styles for visual feedback.

### CODE REVIEW

**Comment by ChrisRenfrow** [comment]
Making a note here that the initial instance of any Thing does not appear to have a UUID. e.g. if I create something and then edit it immediately afterwards, the preceding thing-box will not have a UUID to compare against. Perhaps this could be covered as an edge-case with something like:
```
if (last thing-box name (h1 contents) EQ new thing box name AND last thing box type (npc or place) EQ new thing box type)
OR the UUIDs are present and match:
> do the thing
```

File: core/src/world/npc/view.rs [hunk 1]
```diff
@@ -97,7 +97,13 @@ impl<'a> fmt::Display for DetailsView<'a> {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
         let Self { npc, relations } = self;
 
-        writeln!(f, "<div class=\"thing-box npc\">\n")?;
+        writeln!(
+            f,
+            "<div class=\"thing-box npc\"{}>\n",
+            npc.uuid
+                .clone()
+                .map_or("".to_string(), |v| format!(" data-uuid=\"{}\"", v))
```
Comment (MikkelPaulson) [inline core/src/world/npc/view.rs:105]: I thiiink this is doable with [`format_args!()`](https://doc.rust-lang.org/std/macro.format_args.html), thereby avoiding a redundant allocation.

File: core/src/world/npc/view.rs [hunk 2]
```diff
@@ -97,7 +97,13 @@ impl<'a> fmt::Display for DetailsView<'a> {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
         let Self { npc, relations } = self;
 
-        writeln!(f, "<div class=\"thing-box npc\">\n")?;
+        writeln!(
+            f,
+            "<div class=\"thing-box npc\"{}>\n",
+            npc.uuid
+                .clone()
```
Comment (MikkelPaulson) [inline core/src/world/npc/view.rs:104]: Try with `as_ref()` instead.

**Comment by MikkelPaulson** [comment]
> Making a note here that the initial instance of any Thing does not appear to have a UUID. e.g. if I create something and then edit it immediately afterwards, the preceding thing-box will not have a UUID to compare against.

Oh, the UUID isn't generated until the entity is saved to the journal? If so it seems reasonable to change that logic, making sure that there are no lingering cases like `if thing.uuid.is_some()` as shorthand for "has been saved".

Your proposed logic could run into problems with a command sequence like

```
generate npc
load SomeSavedNpc
```

Where the load command would load the info into the box that was just generated (but not saved).

**Comment by ChrisRenfrow** [comment]
> Where the load command would load the info into the box that was just generated (but not saved).

Perhaps I'm missing something, but according to my proposed logic, wouldn't this only be the case if the loaded npc shares the same name as the generated one? In which case it seems this would trigger a "name already in use" error.

**Comment by MikkelPaulson** [comment]
Ah, good point. Still, it seems a bit fragile when the UUID is right there (or could be easily added).

**Comment by ChrisRenfrow** [comment]
> Oh, the UUID isn't generated until the entity is saved to the journal? If so it seems reasonable to change that logic, making sure that there are no lingering cases like if thing.uuid.is_some() as shorthand for "has been saved".

Am I correct in assuming that this is something that should be handled in a separate PR? Seems like an involved task at a glance but maybe you have a simpler fix in mind.

**Comment by MikkelPaulson** [comment]
Maybe. I'm away from a computer this weekend, but I'll take a look next week. I think it's a simple change but it'll probably touch quite a few files. And it'll make some tests non-deterministic.

File: core/src/world/npc/view.rs (hunk 1 above)
Comment (ChrisRenfrow) [reply core/src/world/npc/view.rs:105]: There may be a more elegant way to express this but this is what I've come up with to avoid the allocation.

```rs
let uuid = npc.uuid.as_ref();

writeln!(
    f,
    "<div class=\"thing-box npc\"{}>\n",
    (uuid.is_some())
        .then_some(format_args!(" data-uuid=\"{}\"", uuid.unwrap()))
        .unwrap_or(format_args!(""))
)?;
```

There is an ownership issue with just dropping in `format_args!()` with the original solution, this manages to get around all of that. Inspired by [this thread](https://internals.rust-lang.org/t/format-args-with-long-lifetimes/19494/6).

Comment (MikkelPaulson) [reply core/src/world/npc/view.rs:105]: In fact, the type will probably change from `Option<Uuid>` to `Uuid` with the UUID-on-create change, rendering all of this moot.

File: core/src/world/npc/view.rs (hunk 2 above)
Comment (ChrisRenfrow) [reply core/src/world/npc/view.rs:104]: I'll be sure to use `as_ref()` where possible. I've gotten perhaps too comfortable with throwing `clone()` everywhere. :)

File: core/src/world/npc/view.rs (hunk 1 above)
Comment (ChrisRenfrow) [reply core/src/world/npc/view.rs:105]: This is true!

**Comment by MikkelPaulson** [comment]
> I think it's a simple change

-- Past @MikkelPaulson

**Comment by ChrisRenfrow** [comment]
No review necessary at the moment as I still need to wrap-up the "diff" code, but feedback is welcome.

**Comment by MikkelPaulson** [comment]
I think it's a bit confusing both from a UX and technical standpoint to collapse edit commands, but the CSS was designed with the expectation of a command-output-command-output cycle. Part of the reason you're running out of space is because consecutive commands have too much vertical margin. A maximum of one line spacing between them seems appropriate, and we could even experiment with no spacing at all.

Maybe it's a result of the UUID weirdness that will be resolved when my branch merges, but my initial testing resulted in two boxes, the latter being the one that was updated when data changed.

Have you experimented with CSS animations? I feel like some sort of flash or fade would better draw attention to the fact that a box is changing contents, which is otherwise not expected behaviour. I'm also curious what impact this will have on screen readers, which are otherwise well supported by the site. That might be an investigation for another issue, though.

Be warned that you're going to run into some challenges with the tutorial. It is trivially broken in your draft PR because it includes some fragile string parsing that was affected by your formatting changes, no big deal to fix. However, I actually have no idea how the "Editing Characters" step is going to interact with these changes. Despite my best efforts, the tutorial is a horrible mass of spaghetti, but at least it is one of the best covered by integration tests, and tends to uncover particularly weird corner cases (eg. #302) since by design it interacts with most of the parts of the application in non-obvious ways.

I do recommend that you wait until the UUID change merges and rebase on my branch before touching the tutorial, since it has a bunch of trivial code changes that will nevertheless leave you buried under a pile of merge conflicts.

**Comment by MikkelPaulson** [comment]
> I do recommend that you wait until the UUID change merges and rebase on my branch before touching the tutorial, since it has a bunch of trivial code changes that will nevertheless leave you buried under a pile of merge conflicts.

Done! Time for a rebase party.
